File: java/src/main/behavioral/command/Invoker.java
package behavioral.command;

import java.util.ArrayList;
import java.util.List;

// Invoker: Maintains a queue of commands and executes them
public class Invoker {
    private final List<Command> commandQueue = new ArrayList<>();

    public void addCommand(Command command) {
        commandQueue.add(command);
    }

    public void executeCommands() {
        for (Command command : commandQueue) {
            command.execute();
        }
        commandQueue.clear();
    }
}

// Example Usage:
// public class Main {
//     public static void main(String[] args) {
//         Receiver receiver = new Receiver();
//         Command command = new ConcreteCommand(receiver);
//
//         Invoker invoker = new Invoker();
//         invoker.addCommand(command);
//
//         // Execute all queued commands
//         invoker.executeCommands();
//
//         // Expected Output:
//         // ConcreteCommand: Executing command.
//         // Receiver: Performing an action.
//     }
// }
//
// Common Mistakes to Avoid:
//
// 1. Storing Too Many Commands in Memory:
//    - Mistake: Keeping executed commands indefinitely in a queue.
//    - Fix: Clear commands after execution if they are no longer needed.
//
// 2. Forgetting to Execute Commands:
//    - Mistake: Adding commands to the invoker but never executing them.
//    - Fix: Ensure `executeCommands()` is called at the appropriate time.
//
// 3. Making Commands Dependent on Specific Receivers:
//    - Mistake: Hardcoding receiver logic inside the command.
//    - Fix: Pass receivers dynamically to allow flexibility.
//
// 4. Overcomplicating Simple Tasks:
//    - Mistake: Using the Command pattern when a simple function reference would suffice.
//    - Fix: Use this pattern only when decoupling commands and receivers is beneficial.

File: python/src/behavioral/command/invoker.py
from typing import Optional
from .command import Command


class Invoker:
    """
    The Invoker stores and triggers commands.
    """

    def __init__(self):
        self._on_start: Optional[Command] = None
        self._on_finish: Optional[Command] = None

    def set_on_start(self, command: Command) -> None:
        self._on_start = command

    def set_on_finish(self, command: Command) -> None:
        self._on_finish = command

    def run(self) -> list[str]:
        output = []

        if self._on_start:
            output.append("Invoker: Starting operation...")
            output.append(self._on_start.execute())

        if self._on_finish:
            output.append("Invoker: Finishing operation...")
            output.append(self._on_finish.execute())

        return output


# Example Usage:
#
# from receiver import Receiver
# from concrete_command_a import ConcreteCommandA
# from concrete_command_b import ConcreteCommandB
# from invoker import Invoker
#
# receiver = Receiver()
# invoker = Invoker()
#
# command_a = ConcreteCommandA(receiver)
# command_b = ConcreteCommandB(receiver)
#
# invoker.set_on_start(command_a)
# invoker.set_on_finish(command_b)
#
# invoker.run()
#
# Output:
# Invoker: Starting operation...
# Receiver: Action A executed.
# Invoker: Finishing operation...
# Receiver: Action B executed.
#
#
# Common Mistakes to Avoid:
#
# 1. Putting business logic inside Commands.
#    - Fix: Keep logic in Receiver, commands should delegate only.
#
# 2. Using too many command classes for trivial operations.
#    - Fix: Use Command Pattern only when you need undo/queue/logging/execution control.
#
# 3. Tight coupling Invoker <-> concrete commands.
#    - Fix: Invoker should work with the Command interface only.
